codefresh/resource_pipeline.go

```diff
@@ -291,93 +282,151 @@ func mapPipelineToResource(cfObject codefreshObject, d *schema.ResourceData) err
-		res = append(res, map[string]interface{}{
-			"name":                trigger.Name,
-			"description":         trigger.Description,
-			"context":             trigger.Context,
-			"repo":                trigger.Repo,
-			"branch_regex":        trigger.BranchRegex,
-			"modified_files_glob": trigger.ModifiedFilesGlob,
-			"disabled":            trigger.Disabled,
-			"provider":            trigger.Provider,
-			"type":                trigger.Type,
-			"events":              trigger.Events,
-			"variables":           convertVariables(trigger.Variables),
-		})
+func flattenTriggers(triggers []cfClient.Trigger) []map[string]interface{} {
+	var res = make([]map[string]interface{}, len(triggers))
+	for i, trigger := range triggers {
+		m := make(map[string]interface{})
+		m["name"] = trigger.Name
+		m["description"] = trigger.Description
+		m["context"] = trigger.Context
+		m["repo"] = trigger.Repo
+		m["branch_regex"] = trigger.BranchRegex
+		m["modified_files_glob"] = trigger.ModifiedFilesGlob
+		m["disabled"] = trigger.Disabled
+		m["provider"] = trigger.Provider
+		m["type"] = trigger.Type
+		m["events"] = trigger.Events
+		m["variables"] = convertVariables(trigger.Variables)
+
+		res[i] = m
 	}
 	return res
 }
 
-func mapResourceToPipeline(d *schema.ResourceData) codefreshObject {
+func mapResourceToPipeline(d *schema.ResourceData) *cfClient.Pipeline {
+
 	tags := d.Get("tags").(*schema.Set).List()
-	pipeline := &pipeline{
-		Metadata: metadata{
-			Name: d.Get("name").(string),
-			Labels: labels{
+
+	pipeline := &cfClient.Pipeline{
+		Metadata: cfClient.Metadata{
+			Name:      d.Get("name").(string),
+			Revision:  d.Get("revision").(int),
+			ProjectId: d.Get("project_id").(string),
+			Labels: cfClient.Labels{
 				Tags: convertStringArr(tags),
 			},
+			OriginalYamlString: strings.Replace(
+				d.Get("original_yaml_string").(string),
+				"\n",
+				"\n",
+				-1),
 		},
-		Spec: &spec{
-			SpecTemplate: &specTemplate{
-				Location: d.Get("spec.0.location").(string),
-				Repo:     d.Get("spec.0.repo").(string),
-				Path:     d.Get("spec.0.path").(string),
-				Revision: d.Get("spec.0.revision").(string),
-				Context:  d.Get("spec.0.context").(string),
-			},
+		Spec: cfClient.Spec{
 			Priority:    d.Get("spec.0.priority").(int),
 			Concurrency: d.Get("spec.0.concurrency").(int),
 		},
 	}
-	variables := d.Get("variables").(map[string]interface{})
-	pipeline.setVariables(variables)
 
-	triggers := d.Get("trigger").([]interface{})
+	if _, ok := d.GetOk("spec.0.spec_template"); ok {
+		pipeline.Spec.SpecTemplate = &cfClient.SpecTemplate{
+			Location: d.Get("spec.0.spec_template.0.location").(string),
+			Repo:     d.Get("spec.0.spec_template.0.repo").(string),
+			Path:     d.Get("spec.0.spec_template.0.path").(string),
+			Revision: d.Get("spec.0.spec_template.0.revision").(string),
+			Context:  d.Get("spec.0.spec_template.0.context").(string),
+		}
+	}
+
+	if _, ok := d.GetOk("spec.0.runtime_environment"); ok {
```

Comment:
what is "ok" why not "err"?

Reply:
```
// GetOk returns the data for the given key and whether or not the key
// has been set to a non-zero value at some point.
//
// The first result will not necessarilly be nil if the value doesn't exist.
// The second result should be checked to determine this information.
func (d *ResourceData) GetOk(key string) (interface{}, bool) {
	r := d.getRaw(key, getSourceSet)
	exists := r.Exists && !r.Computed
	if exists {
		// If it exists, we also want to verify it is not the zero-value.
		value := r.Value
		zero := r.Schema.Type.Zero()

		if eq, ok := value.(Equal); ok {
			exists = !eq.Equal(zero)
		} else {
			exists = !reflect.DeepEqual(value, zero)
		}
	}

	return r.Value, exists
}
```

I think `ok` is a better name for this var.
`if <ok=this parameter is set> { do something }`